Stat

bretambrose opened the pull request:
Enum range calculations and package id
Socket channel creation options

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
WIP

bretambrose committed:
WIP

Commit:
CR Feedback

Commit:
Merge branch 'Stat' into Stat2

Commit:
WIP

bretambrose committed:
Merge branch 'master' into Stat

Commit:
WIP

Commit:
WIP

Commit:
WIP

Commit:
WIP

Commit:
Tests passing

Commit:
Doc

Commit:
Merge branch 'Stat' of github.com:awslabs/aws-c-http into Stat

Commit:
Dummy change to trigger CI

Commit:
Uninited value

bretambrose committed:
Merge branch 'master' into Stat

Commit:
tidy

Commit:
Merge branch 'Stat' of github.com:awslabs/aws-c-http into Stat

Commit:
unreferenced

Commit:
Remove unused error:

Commit:
remove atomic

Commit:
conversion

Commit:
Proper format specifiers

bretambrose committed:
Merge branch 'master' into Stat

bretambrose committed:
Merge branch 'master' into Stat

Commit:
Fix http stats tests

graebm commented on source/connection_monitor.c:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    return options->minimum_throughput_failure_threshold_in_seconds > 1 &&
+           options->minimum_throughput_bytes_per_second > 0;
```
The comments warn that 1 might result in flaky behavior, but here you outright refuse to do monitoring if they use it

bretambrose replied:
Modified the name and meaning a bit.

graebm commented on source/connection_monitor.c:
```diff
@@ -0,0 +1,200 @@
+        impl->consecutive_throughput_failures);
+
+    if (impl->consecutive_throughput_failures < impl->options.minimum_throughput_failure_threshold_in_seconds) {
+        return;
```
Seems like it we'd get better behavior if we used the actual time elapsed between ticks.

Kind of like how games **try** to tick at 60FPS, but you get weird behavior if you write your game code assuming it's running at a constant tick-rate, instead of coding against actual `float delta`

instead of just counting 1,2,3 each time one of these intervals failed, you could sum the length of time they'd been failing for. So if you failed on the first tick, but you were only reading/writing a fraction of that time, then you wouldn't exceed the `minimum_throughput_failure_threshold_in_sec` because you'd only been failing for, say, 100ms so far.

graebm replied:
An approach like this might make the issues you were seeing with 1 second thresholds go away.

bretambrose replied:
I dunno, I'd rather force multiple samples unconditionally.  Using actual time elapsed, any single long hiccup (1 bad callback for example) will shut down the connection.  In my update I've renamed the config parameter away from a notion of time.

bretambrose replied:
Entire approach reworked

graebm commented on source/connection_monitor.c:
```diff
@@ -0,0 +1,200 @@
+        return;
+    }
+
+    bytes_per_second /= pending_io_time_ns;
```
maybe use `double` or `float` in these calculations so we don't have to worry about integer math chopping off remainders coming up with fishy numbers

bretambrose replied:
Done.

graebm commented on source/connection_monitor.c:
```diff
@@ -0,0 +1,200 @@
+    if (impl->options.minimum_throughput_bytes_per_second == 0 ||
+        impl->options.minimum_throughput_failure_threshold_in_seconds == 0) {
+        return;
+    }
```
[trivial] this could be moved earlier in the fn to avoid unnecessary work

graebm commented on source/connection_monitor.c:
```diff
@@ -0,0 +1,200 @@
+    impl->consecutive_throughput_failures = 0;
+
+    uint64_t pending_io_time_ns = 0;
+    if (aws_add_u64_checked(pending_read_interval_ns, pending_write_interval_ns, &pending_io_time_ns)) {
```
I don't understand why you're adding together the read and write intervals. If there's a stream that's been sending the entire time, you'll end up with a `pending_io_time = 2 seconds`
and a `total_bytes = N sent bytes` which will get you `bytes_per_second = N/2`

I would think you'd want to report `bytes_per_second = N`

Maybe it would be better to calculate the `read_bytes_per_second` and `write_bytes_per_second` separately, and either ADD those or take the MAX of the two?

bretambrose replied:
That was just dumb on my part.

Commit:
Fix dumb mistakes

bretambrose commented on include/aws/http/connection.h:
```diff
@@ -56,6 +56,25 @@ typedef void(
+     * will be performed.  A value of two or greater is STRONGLY recommended.  A value of one will result in
+     * timing-dependent, unreliable behavior.
+     */
+    uint32_t minimum_throughput_failure_threshold_in_seconds;
```
Reworked parameter meaning away from a notion of time.

Commit:
WIP

Commit:
WIP

Commit:
New approach

Commit:
Merge branch 'master' into Stat

Commit:
tidy/misc fixes

Commit:
Merge branch 'master' into Stat

JonathanHenson commented on source/connection_monitor.c:
```diff
@@ -0,0 +1,217 @@
+        (void *)channel,
+        bytes_per_second);
+
+    /* Check throughput only if at least one stream exists and was observed in that role pregviously */
```
typo, pregviously

JonathanHenson commented on source/connection_monitor.c:
```diff
@@ -0,0 +1,217 @@
+        (current_incoming_stream_id != 0 && current_incoming_stream_id == impl->last_incoming_stream_id) ||
+        (current_outgoing_stream_id != 0 && current_outgoing_stream_id == impl->last_outgoing_stream_id);
+
+    impl->last_outgoing_stream_id = current_outgoing_stream_id;
```
how does this handle parallel streams? wouldn't we need a monitor per stream in that case?

bretambrose replied:
When we update this to support H2, this particular logic would only apply to h1 streams.  H2 streams would have their own stats structure which we can configure to contain the data we need to make a similar condition (if that makes sense) or for h2 streams we could always just drop this requirement.  

The fact that the streams have a simple monotonic id mechanism could mean that just keeping track of the current min and max within a connection is sufficient to apply the same conditional filter (ie if the current min id is less than the previous max id then you know at least one stream exists from the previous sample timepoint).

bretambrose committed:
Update

graebm commented on source/h1_connection.c:
```diff
@@ -615,6 +627,50 @@ static void s_stream_complete(struct aws_h1_stream *stream, int error_code) {
+}
+
+static void s_set_outgoing_stream_ptr(struct h1_connection *connection, struct aws_h1_stream *next_outgoing_stream) {
+    struct aws_h1_stream *prev = connection->thread_data.outgoing_stream;
```
Just a heads-up that when we change the outgoing_stream_ptr, it doesn't necessarily mean that the stream's data is done writing to the socket. It just means we've copied the stream's last outgoing bytes to a `aws_channel_message`. So for a stream whose outgoing data all fits in 16KB, it's only going to be the "current outgoing stream" for a fleeting moment, even if the socket ends up taking minutes to actually write that full `aws_channel_message`

Maybe what we want to measure is the time from when we shipped the `aws_channel_message` down the channel, to when it finished writing (there is a callback when the write completes, but the H1 code isn't currently hooked into it)

bretambrose replied:
That would be a precision improvement, but how complicated would its addition be?  Seems like something that could wait.

graebm commented on source/connection_monitor.c:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    AWS_ZERO_STRUCT(*handler);
+    AWS_ZERO_STRUCT(*impl);
```
style/opinion: you could just embed the `aws_crt_statistics_handler` within the impl struct. Then you can just use one calloc instead of acquire_many and 2 ZERO_STRUCTS

bretambrose replied:
Tbh I find that approach confusing.

graebm commented on source/connection_monitor.c:
```diff
@@ -0,0 +1,222 @@
+        if (fractional_bytes_written_per_second >= (double)UINT64_MAX) {
+            bytes_per_second = UINT64_MAX;
+        } else {
+            bytes_per_second = (uint64_t)fractional_bytes_written_per_second;
```
I think you can just do `bytes_per_second = (uint64_t)fractional_bytes_written_per_second;`
I played around in a toy program and `((uint64_t)DBL_MAX == UINT64_MAX)` was true

bretambrose replied:
Digging further, it appears that casting a too-large double to an integer type is undefined behavior:

6.3.1.4 Real floating and integer

When a finite value of real ﬂoating type is converted to an integer type other than _Bool, the fractional part is discarded (i.e., the value is truncated toward zero). If the value of the integral part cannot be represented by the integer type, the behavior is undefined.


From http://blog.frama-c.com/index.php?post/2013/10/09/Overflow-float-integer

graebm commented on source/connection_monitor.c:
```diff
@@ -0,0 +1,222 @@
+
+    if (pending_write_interval_ms > 0) {
+        double fractional_bytes_written_per_second =
+            (double)bytes_written * (double)AWS_TIMESTAMP_MILLIS / (double)pending_write_interval_ms;
```
I think you should replace `AWS_TIMESTAMP_MILLIS` here with MY_REPORT_INTERVAL_MS?

`AWS_TIMESTAMP_MILLIS` happens to be 1000, and `s_get_report_interval_ms()` has `return 1000;`
but using the same constant/define in both places would be preferable.

Assuming my understanding of this math is correct

Or should we measure the ACTUAL time elapsed since the last sample and use that here, instead of a hardcoded number?

bretambrose replied:
We're using the actual time we were "trying" to write (with the caveat that, as you pointed out elsewhere, it's not completely accurate to the socket level).  The stats reporting interval is completely independent and not a factor in this calculation.

graebm commented on source/connection_monitor.c:
```diff
@@ -0,0 +1,222 @@
+        if (fractional_bytes_read_per_second >= (double)UINT64_MAX) {
+            bytes_per_second = UINT64_MAX;
+        } else {
+            bytes_per_second = aws_add_u64_saturating(bytes_per_second, (uint64_t)fractional_bytes_read_per_second);
```
trivial: not sure if bytes_per_second should be a SUM or a MAX of the read and write
either works for the purposes of killing a slow connection
but I don't know who else might be using stats and how they might see it

bretambrose replied:
Well this particular calculation result isn't consumable on its own, so nothing else can use it atm.

bretambrose commented on source/connection.c:
```diff
@@ -627,6 +633,17 @@ static void s_client_bootstrap_on_channel_setup(
+    if (aws_http_connection_monitoring_options_is_valid(&http_bootstrap->monitoring_options)) {
+        struct aws_crt_statistics_handler *http_connection_monitor =
+            aws_crt_statistics_handler_new_http_connection_monitor(
+                http_bootstrap->alloc, &http_bootstrap->monitoring_options);
```
Looks like I'm conflating invalid and not-set.  So maybe creation should check for set-and-invalid and log that as an error and continue with not-set.  Everywhere else then assumes not-set or valid.

bretambrose replied:
Looking at it closer, I think I'll just track monitoring options by pointer and thus we can assume they're valid after creation time, if non-null.  Will have to cleanup on channel shutdown when the http bootstrap is freed.

bretambrose commented on source/h1_connection.c:
```diff
@@ -615,6 +627,50 @@ static void s_stream_complete(struct aws_h1_stream *stream, int error_code) {
+        /* transition from nothing to write -> something to write */
+        connection->thread_data.outgoing_stream_timestamp_ns = now_ns;
+    } else if (prev != NULL && next_outgoing_stream == NULL) {
+        /* transition from something to write -> nothing to write */
```
This logic is just "start measuring"/"stop measuring"  Transitioning streams should have no effect.

bretambrose commented on source/h1_connection.c:
```diff
@@ -1751,3 +1808,51 @@ static size_t s_handler_message_overhead(struct aws_channel_handler *handler) {
+            &connection->thread_data.stats.pending_outgoing_stream_ms);
+        connection->thread_data.outgoing_stream_timestamp_ns = now_ns;
+        connection->thread_data.stats.current_outgoing_stream_id =
+            aws_http_stream_get_id(&connection->thread_data.outgoing_stream->base);
```
I blame clang format.  Was all one long line before!

bretambrose commented on include/aws/http/statistics.h:
```diff
@@ -0,0 +1,63 @@
+ * A statistics struct for http handlers.  Tracks the actual amount of time that incoming and outgoing requests are
+ * waiting for their IO to complete.
+ */
+struct aws_crt_statistics_http1_channel {
```
Leaving it public allows people to write their own monitors.

Commit:
More PR feedback

bretambrose commented on source/h1_connection.c:
```diff
@@ -1751,3 +1808,51 @@ static size_t s_handler_message_overhead(struct aws_channel_handler *handler) {
+            &connection->thread_data.stats.pending_incoming_stream_ms);
+        connection->thread_data.incoming_stream_timestamp_ns = now_ns;
+        connection->thread_data.stats.current_incoming_stream_id =
+            aws_http_stream_get_id(&connection->thread_data.incoming_stream->base);
```
We do that right there (in the middle of the hard-to-read part).

bretambrose committed:
Merge branch 'master' into Stat

Commit:
Merge fixes

Commit:
Upgrade to clang-9

Commit:
Windows warning

bretambrose committed:
Merge branch 'master' into Stat

graebm committed:
Control rate of HTTP/1 writes (#180)

* HTTP/1 only writes 1 message at a time.
This avoids bloating memory with unsent messages.
This also helps improve the accuracy of our upload stats.
Still needs tweaking to deal with websocket upgrade edge-cases.

* Install websocket handler the moment 101 response comes in, do not wait for stream to complete.
Remove hacks where 101 response completed a stream, no 1xx response should complete a stream.

* Simplify confusing list logic.
Remove the server-only `waiting_stream_list`.
Do a simple scan over available streams to find the next one to work on. This is O(N) when pipelining, but I hear that no one does HTTP/1 pipelining, so in that case O(N=1). We can always re-optimize if we have data that pipelining is super popular and this is a hotspot.

Commit:
Update

bretambrose committed:
Merge branch 'master' into Stat

Commit:
Windows warning
